Sources/NIOSSL/SubjectAlternativeName.swift

```diff
@@ -0,0 +1,201 @@
+        }
+
+        deinit {
+            CNIOBoringSSL_GENERAL_NAMES_free(self.nameStack)
+        }
+    }
+    
+    
+    @usableFromInline internal var storage: Storage
+    
+    internal init(nameStack: OpaquePointer) {
+        self.storage = .init(nameStack: nameStack)
+    }
+}
+
+extension _SubjectAlternativeNames: RandomAccessCollection {
+    
+    @inlinable public subscript(position: Int) -> _SubjectAlternativeName {
+        self.storage[position]
+    }
+    
+    @inlinable public var startIndex: Int { 0 }
+    @inlinable public var endIndex: Int { self.storage.stackSize }
+}
+
+public struct _SubjectAlternativeName {
+    
+    public struct NameType: Hashable {
+        public var rawValue: Int
+
+        public init(_ rawCode: Int) {
+            self.rawValue = rawCode
+        }
+
+        fileprivate init(_ rawCode: Int32) {
+            self.init(Int(rawCode))
+        }
+        
+        public static let email = Self(GEN_EMAIL)
+        public static let dnsName = Self(GEN_DNS)
+        public static let ipAddress = Self(GEN_IPADD)
+    }
+
+    public struct Contents {
+        // only part of this type to keep a strong reference to the underlying storage of `buffer`
+        private let collection: _SubjectAlternativeNames.Storage
+        // lifetime automatically managed by `collection`
+        @usableFromInline internal let buffer: UnsafeBufferPointer<UInt8>
+        
+        internal init(collection: _SubjectAlternativeNames.Storage, buffer: UnsafeBufferPointer<UInt8>) {
+            self.collection = collection
+            self.buffer = buffer
+        }
+        
+        @inlinable public func withUnsafeBufferPointer<Result>(
+            _ body: (UnsafeBufferPointer<UInt8>) throws -> Result
+        ) rethrows -> Result {
+            try body(self.buffer)
+        }
+    }
+    
+    // should be replaced by `swift-nio`s `IPAddress` once https://github.com/apple/swift-nio/issues/1650 is resolved
+    internal enum IPAddress {
+        case ipv4(in_addr)
+        case ipv6(in6_addr)
+    }
+    
+    public var nameType: NameType
+    public var contents: Contents
+}
+
+extension _SubjectAlternativeName.Contents: RandomAccessCollection {
+    
+    @inlinable public var startIndex: Int { self.buffer.startIndex }
+    @inlinable public var endIndex: Int { self.buffer.endIndex }
+    
+    @inlinable public subscript(position: Int) -> UInt8 {
+        self.buffer[position]
+    }
+}
+
+extension _SubjectAlternativeName.IPAddress {
+    
+    internal init?(_ subjectAlternativeName: _SubjectAlternativeName) {
+        guard subjectAlternativeName.nameType == .ipAddress else {
+            return nil
+        }
+        switch subjectAlternativeName.contents.count {
+        case 4:
+            let addr = subjectAlternativeName.contents.withUnsafeBufferPointer {
+                $0.baseAddress?.withMemoryRebound(to: in_addr.self, capacity: 1) {
```

Comment:
Oh we just cleaned this up in NIO, let's do the same here: this should be:

```
$0.baseAddress.map { UnsafeRawPointer($0) }.load(in_addr.self)
```

The same transformation should go on below as well.